### ISSUE TITLE
Network timeout when posting large media files

### ISSUE THREAD

**Issue by MastodonContentMover** [andregasser/bigbone#197]
**Problem:**

_MastodonClient.kt_ currently experiences a _SocketTimeoutException_ when uploading large media files to Mastodon instances, apparently originating within the OkHttp3 library used for HTTP connections.

Here's the stack trace produced today when uploading a 37.4MB mp4 video file to the mas.to instance:
![timeout](https://user-images.githubusercontent.com/129753354/234603483-1d352830-fd97-4cc6-bc9f-55475f1cbcec.JPG)

During reattempts, timeouts occurred after varying intervals (sometimes around 45 seconds, sometimes up to around 70 seconds). The file involved had successfully been uploaded to Mastodon instances without issue. This and the fact that it later (see below) uploaded successfully suggests the problem was not specific to the file (resolution, format etc.)

**Workaround:**

Although the operation where this is seen involves uploading data to the instance, setting _useStreamingApi()_ when creating the _MastodonClient_ object (which affects only the read timeout) appears to help and the exception has not yet been seen with this setting. It's unclear why a read timeout would affect a write operation (presumably the issue with length of time taken to transfer data relates to the size of the media file — the issue hasn't been seen with smaller videos or photos). 

**Possible solutions:**

It may be useful to offer BigBone library users the ability to specify a timeout when initializing their _MastodonClient_ instance, or to internally set the affected timeouts to a higher value for operations involving large files likely to take an extended period to transfer. 

@bocops advises that there are plans to use the newer asynchronous Mastodon API for media uploads in the future, and there may be limited value in adjusting the existing code if elaborate work is needed to lengthen the timeout or allow users to customize their _MastodonClient_ timeouts. Nonetheless, relying on _useStreamingApi()_ also isn't ideal as it sounds likely that the function of this method will change in the future, which may have unintended consequences for anyone using it purely to adjust a timeout.

**In summary:**

It looks like I've been able to work around this using _useStreamingApi()_ (I haven't tested this with lots of videos but it's unlikely I'll be trying to upload any much larger than 37.4MB), but it would be good to determine a more futureproof fix or workaround if possible.

**Comment by bocops** [comment]
I guess it lies in the nature of HTTP connection timeouts that they are hard to make sense of. Even temporary issues, whether it is server load, your own device or something inbetween, could lead to the same call failing one time but being successful the next.

I have no idea why increasing a **read** timeout helps a call that seems to restricted mostly by the amount of data it **writes** - but it seems sensible to allow library users to decide for themselves what timeout values they need.

The PR I just sent introduces three new methods in the builder that control _read_, _write_ and _connect_ timeouts respectively. I also changed the existing `useStreamingApi()` method to not interfere if the read timeout has been set to a higher value before calling it.

**Comment by MastodonContentMover** [comment]
That sounds like it would take care of this very well - thank you, and thank you for looking at it so quickly!

**Comment by andregasser** [comment]
@bocops Did you already open the PR? Can't find it (on my mobile).

**Comment by bocops** [comment]
@andregasser Sorry, my fault. There's a PR now. :)
